### PR TITLE
[SPARK-36524][SQL] Common class for ANSI interval types

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Add new type `AnsiIntervalType` to `AbstractDataType.scala`, and extend it by `YearMonthIntervalType` and by `DayTimeIntervalType`

### Why are the changes needed?
To improve code maintenance. The change will allow to replace checking of both `YearMonthIntervalType` and `DayTimeIntervalType` by a check of `AnsiIntervalType`, for instance:
```scala
    case _: YearMonthIntervalType | _: DayTimeIntervalType => false
``` 
by
```scala
    case _: AnsiIntervalType => false
```

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By existing test suites.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@gengliangwang @sarutak @cloud-fan @AngersZhuuuu @beliefer @Peng-Lei Could you review this PR, please.

**Review by gengliangwang — APPROVED** [review]
+1

**Review by sarutak — APPROVED** [review]
+1, Pending CIs.

**Review by AngersZhuuuu — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala [hunk 1]
```diff
@@ -222,3 +222,11 @@ private[sql] object AnyTimestampType extends AbstractDataType with Serializable
 
   def unapply(e: Expression): Boolean = acceptsType(e.dataType)
 }
+
+/**
+ * The interval type which conforms to the ANSI SQL standard.
+ *
+ * @since 3.2.0
+ */
+@Unstable
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala:231]: do we need this annotation? it's `private[sql]` and not a public API

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala:231]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala [hunk 2]
```diff
@@ -222,3 +222,11 @@ private[sql] object AnyTimestampType extends AbstractDataType with Serializable
...
+private[sql] abstract class AnsiIntervalType extends AtomicType
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala:232]: Is it correct to say `AtomicType`? `CalendarIntervalType` extends `DataType`.

**Review by HyukjinKwon — APPROVED** [review]

**Review by beliefer — APPROVED** [review]
+1

**Comment by Peng-Lei** [comment]
+1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala:231]: ok. Let me remove `@Unstable`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala [hunk 3]
```diff
@@ -222,3 +222,10 @@ private[sql] object AnyTimestampType extends AbstractDataType with Serializable
 
   def unapply(e: Expression): Boolean = acceptsType(e.dataType)
 }
+
+/**
+ * The interval type which conforms to the ANSI SQL standard.
+ *
+ * @since 3.2.0
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala:229]: Let's also remove these since this isn't an API.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by MaxGekk** [comment]
Thanks to all. Merging to master/3.2.

**Pull request closed without merging**